lib/charms/blackbox_k8s/v0/blackbox_probes.py

```diff
@@ -0,0 +1,634 @@
+        modules: Optional[Dict] = None,
+        refresh_event: Optional[Union[BoundEvent, List[BoundEvent]]] = None,
+        relation_name: str = DEFAULT_RELATION_NAME,
+    ):
+        super().__init__(charm, relation_name)
+        """Construct a Blackbox Exporter client.
+
+        Charms seeking to expose metric endpoints to be probed via Blackbox, must do so
+        using the `BlackboxProbesProvider` object from this charm library.
+        For the simplest use cases, the BlackboxProbesProvider object requires
+        to be instantiated with a list of jobs with the endpoints to monitor.
+        A probe in blackbox is defined by a module and a static_config target. Those
+        are then organised in a prometheus job for proper scraping.
+        The `BlackboxProbesProvider` constructor requires
+        the name of the relation over which a probe target
+        is exposed to the Blakcbox Exporter charm.
+
+        Args:
+            charm: a `CharmBase` object which manages the
+                `BlackboxProbesProvider` object. Generally, this is `self` in
+                the instantiating class.
+            probes: the probes to configure in Blackbox Exporter passed as a list
+                    of configuration probes in python structures.
+            modules: an optional definition of modules for Blackbox Exporter to
+                use. For details on how to write pass a module, see the
+                [docs upstream](https://github.com/prometheus/blackbox_exporter/blob/master/CONFIGURATION.md).
+                Further examples are provided [upstream](https://github.com/prometheus/blackbox_exporter/blob/master/example.yml).
+            refresh_event: additional `CharmEvents` event (or a list of them)
+                on which the probes and modules should be updated.
+            relation_name: name of the relation providing the Blackbox Probes
+                service. It's recommended to not change it, to ensure a
+                consistent experience across all charms that use the library.
+        """
+        self.topology = JujuTopology.from_charm(charm)
+        self._charm = charm
+        self._relation_name = relation_name
+        self._probes = [] if probes is None else probes
+        self._modules = {} if modules is None else modules
+
+        events = self._charm.on[self._relation_name]
+        self.framework.observe(events.relation_changed, self._set_probes_spec)
+
+        refresh_event = []
+        if not refresh_event:
+            if len(self._charm.meta.containers) == 1:
+                container = list(self._charm.meta.containers.values())[0]
+                refresh_event = [self._charm.on[container.name.replace("-", "_")].pebble_ready]
+        elif not isinstance(refresh_event, list):
+            refresh_event = [refresh_event]
+
+        module_name_prefix = f"juju_{self.topology.identifier}_"
+        self._prefix_probes(module_name_prefix)
+        self._prefix_modules(module_name_prefix)
```

Comment:
I guess it depends how unique is the topology identifier, I relied on the library documentation that mentions `Juju topology is to uniquely identify a piece of software running across any of your Juju-managed deployments.`

Reply:
This is now addressed on the requirer side (function called `_process_and_hash_probes`),  the probes job_name are all hashed and the ones with same hash (identical probes) discarded. In this way, we avoid collision also between file probes and relation probes in the charm during mergin. Decided to hash all of the probes because we could expect to have thousands of devices, and it feels more efficient than following the dedupe algorithm in prometheus_scrape library,